### ISSUE TITLE
/me/purchases times out for users with thousands of purchases

### ISSUE THREAD

**Issue by dsas** [Automattic/wp-calypso#96639]
### Quick summary

See the parent issue for more information

/me/purchases (**NOTE:** now calypso uses `/upgrades` for this data since [SHILL-1115](https://linear.app/a8c/issue/SHILL-1115/should-we-standardize-on-upgrades-namespace-for-subscription-endpoints)) returns a HTTP 504 - gateway timeout for users with \~3000 subs. This has been noticed on /domains/manage but will probably cause them problems on other calypso pages too.

This happens for at least one user - it sounds like there have been repeated problems with this user's account for example p1694033164744309-slack-C096PD42U

### Steps to reproduce

1. Switch to the user mentioned in the first message here: p1732188303637839-slack-C07GZ2UA3TN
2. Open up developer tools and switch to the network tab
3. Go to /domains/manage
4. In the network tab, filter for purchases and wait

### What you expected to happen

A bunch of results

### What actually happened

A http 504

### Impact

One

### Available workarounds?

No and the platform is unusable

### If the above answer is "Yes...", outline the workaround.

*No response*

### Platform (Simple and/or Atomic)

*No response*

### Logs or notes

*No response*

**Comment by sirbrillig** [comment]
Yes. This is a known issue: https://github.com/Automattic/payments-shilling/issues/2199

Many users will see an improvement when https://github.com/Automattic/payments-shilling/issues/3173 merges but it doesn't look like those performance improvements will help the specific user in the original issue. We'll need to make further improvements to that endpoint.

**Comment by michaeldcain** [comment]
Given the low number of users with thousands of purchases and Shilling's current workload, I'm inclined to lower the priority on this issue. Since we're still discussing it, I'm going to return the "Needs Triage" label.

**Comment by jartes** [comment]
👋🏻 @michaeldcain Seeing that your team will work on this one, and we have a clear picture of this issue, can we mark this one as triaged? 

I'm lowering the priority as you suggested since we are not getting many user reports for this one (and also to accommodate your backlog). 

Thank you 😄

**Comment by Robertght** [comment]
I'm marking this as Triaged on our end, but please ping us again if you need our input.
